File: js/src/keys.ts
/** Per-key runtime state: breaker + rpm bucket + daily quota + cooldowns. */
import { CircuitBreaker } from "./breaker.js";
import { TokenBucket } from "./ratelimit.js";

export const DAY = 86400.0;
/** Stand-in for "unlimited" daily quota so it ranks high but stays finite/comparable. */
export const UNLIMITED = 100_000.0;

export class KeyState {
  breaker = new CircuitBreaker();
  bucket: TokenBucket | null = null;
  rpd: number | null = null;
  rpdUsed = 0;
  rpdReset = 0;
  cooldownUntil = 0;
  disabled = false;
  ewmaLatency = 0;
  lastError: string | null = null;

  constructor(public key: string, public tier = "free") {}

  private rollDaily(now: number): void {
    if (this.rpd === null) return;
    if (this.rpdReset === 0) this.rpdReset = now + DAY;
    else if (now >= this.rpdReset) {
      this.rpdUsed = 0;
      this.rpdReset = now + DAY;
    }
  }

  ready(now: number): boolean {
    if (this.disabled) return false;
    if (now < this.cooldownUntil) return false;
    if (!this.breaker.allow(now)) return false;
    this.rollDaily(now);
    if (this.rpd !== null && this.rpdUsed >= this.rpd) return false;
    if (this.bucket && this.bucket.peek(now) < 1) return false;
    return true;
  }

  reserve(now: number): boolean {
    this.rollDaily(now);
    if (this.bucket && !this.bucket.consume(1, now)) return false;
    this.rpdUsed++;
    return true;
  }

  remaining(now: number): number {
    if (!this.ready(now)) return 0;
    const daily = this.rpd === null ? UNLIMITED : Math.max(0, this.rpd - this.rpdUsed);
    const burst = this.bucket ? this.bucket.peek(now) : UNLIMITED;
    return Math.min(daily, burst);
  }

  waitTime(now: number): number | null {
    if (this.disabled) return null;
    const waits: number[] = [];
    if (now < this.cooldownUntil) waits.push(this.cooldownUntil - now);
    waits.push(this.breaker.timeUntilHalfOpen(now));
    this.rollDaily(now);
    if (this.rpd !== null && this.rpdUsed >= this.rpd) waits.push(Math.max(0, this.rpdReset - now));
    if (this.bucket && this.bucket.peek(now) < 1) waits.push(this.bucket.timeUntil(1, now));
    return waits.length ? Math.max(...waits) : 0;
  }

  masked(): string {
    const k = this.key;
    return k.length > 12 ? `${k.slice(0, 6)}...${k.slice(-4)}` : "***";
  }
}

export function newKeyState(key: string, tier: string, rpm: number | null, rpd: number | null): KeyState {
  const ks = new KeyState(key, tier);
  ks.bucket = rpm ? new TokenBucket(rpm) : null;
  ks.rpd = rpd;
  return ks;
}

File: src/freelm/_keys.py
"""Per-key runtime state: breaker + rpm bucket + daily quota + cooldowns."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from ._breaker import CircuitBreaker
from ._ratelimit import TokenBucket

DAY = 86400.0


@dataclass
class KeyState:
    key: str
    tier: str = "free"
    breaker: CircuitBreaker = field(default_factory=CircuitBreaker)
    bucket: Optional[TokenBucket] = None
    rpd: Optional[int] = None           # requests-per-day cap (None = unknown/unlimited)
    rpd_used: int = 0
    rpd_reset: float = 0.0              # monotonic ts at which the daily counter rolls
    cooldown_until: float = 0.0
    disabled: bool = False              # hard-off after auth failure
    ewma_latency: float = 0.0
    last_error: Optional[str] = None

    # -- daily window ----------------------------------------------------
    def _roll_daily(self, now: float) -> None:
        if self.rpd is None:
            return
        if self.rpd_reset == 0.0:
            self.rpd_reset = now + DAY
        elif now >= self.rpd_reset:
            self.rpd_used = 0
            self.rpd_reset = now + DAY

    # -- gating ----------------------------------------------------------
    def ready(self, now: float) -> bool:
        if self.disabled:
            return False
        if now < self.cooldown_until:
            return False
        if not self.breaker.allow(now):
            return False
        self._roll_daily(now)
        if self.rpd is not None and self.rpd_used >= self.rpd:
            return False
        if self.bucket is not None and self.bucket.peek(now) < 1:
            return False
        return True

    def reserve(self, now: float) -> bool:
        """Consume one rpm token + one daily slot just before firing a request."""
        self._roll_daily(now)
        if self.bucket is not None and not self.bucket.consume(1, now):
            return False
        self.rpd_used += 1
        return True

    def remaining(self, now: float) -> float:
        """A rough 'how much headroom' score for quota-aware routing."""
        self._roll_daily(now)
        daily = float("inf") if self.rpd is None else float(max(0, self.rpd - self.rpd_used))
        burst = self.bucket.peek(now) if self.bucket is not None else float("inf")
        return min(daily, burst)

    def wait_time(self, now: float) -> Optional[float]:
        """Seconds until this key could be ready again, or None if permanently off."""
        if self.disabled:
            return None
        waits = []
        if now < self.cooldown_until:
            waits.append(self.cooldown_until - now)
        waits.append(self.breaker.time_until_half_open(now))
        self._roll_daily(now)
        if self.rpd is not None and self.rpd_used >= self.rpd:
            waits.append(max(0.0, self.rpd_reset - now))
        if self.bucket is not None and self.bucket.peek(now) < 1:
            waits.append(self.bucket.time_until(1, now))
        return max(waits) if waits else 0.0

    def masked(self) -> str:
        k = self.key
        return (k[:6] + "..." + k[-4:]) if len(k) > 12 else "***"


def new_key_state(key: str, *, tier: str, rpm: Optional[float], rpd: Optional[int]) -> KeyState:
    return KeyState(
        key=key,
        tier=tier,
        breaker=CircuitBreaker(),
        bucket=TokenBucket(rate_per_min=rpm) if rpm else None,
        rpd=rpd,
    )
